packages/aws-cdk/lib/api/hotswap/s3-bucket-deployments.ts

```diff
@@ -0,0 +1,125 @@
+import { ISDK } from '../aws-auth';
+import { ChangeHotswapImpact, ChangeHotswapResult, HotswapOperation, HotswappableChangeCandidate, establishResourcePhysicalName } from './common';
+import { EvaluateCloudFormationTemplate } from './evaluate-cloudformation-template';
+
+/**
+ * This means that the value is required to exist by CloudFormation's API (or our S3 Bucket Deployment Lambda)
+ * but the actual value specified is irrelevant
+ */
+export const REQUIRED_BY_CFN = 'required-to-be-present-by-cfn';
+
+export async function isHotswappableS3BucketDeploymentChange(
+  logicalId: string, change: HotswappableChangeCandidate, evaluateCfnTemplate: EvaluateCloudFormationTemplate,
+): Promise<ChangeHotswapResult> {
+  // In old-style synthesis, the policy used by the lambda to copy assets Ref's the assets directly,
+  // meaning that the changes made to the Policy are artifacts that can be safely ignored
+  if (change.newValue.Type === 'AWS::IAM::Policy') {
+    return changeIsForS3DeployCustomResourcePolicy(logicalId, change, evaluateCfnTemplate);
+  }
+
+  if (change.newValue.Type !== 'Custom::CDKBucketDeployment') {
+    return ChangeHotswapImpact.REQUIRES_FULL_DEPLOYMENT;
+  }
+
+  // note that this gives the ARN of the lambda, not the name. This is fine though, the invoke() sdk call will take either
+  const functionName = await establishResourcePhysicalName(logicalId, change.newValue.Properties?.ServiceToken, evaluateCfnTemplate);
```

Comment:
This is wrong I'm pretty sure.

The `logicalId` is for the resource `Custom::CDKBucketDeployment`, _not_ for the Lambda function. Which means this will not actually fall back correctly if `change.newValue.Properties?.ServiceToken` fails evaluation in the template.

This probably means we're missing a unit test for this case.